### PR TITLE
[SYCL] Implement braced-init-list or a number as range for queue::parallel_for

### PR DESCRIPTION
**Pull request by rarutyun**
Modification
- Make three overloads for `queue::parallel_for` to support `range` implicit conversion from number or `braced-init-list`
- Add tests for `queue::parallel_for` calls with generic lambda

Implement the following `queue` [extension](https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/QueueShortcuts/)

Signed-off-by: Ruslan Arutyunyan <ruslan.arutyunyan@intel.com>

### CODE REVIEW

File: sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// UNSUPPORTED: cuda
+// CUDA does not support unnamed lambdas.
```
Comment (bader) [inline sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp:2]: Is using "unnamed lambdas" necessary for this test queue API?

File: sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#include <type_traits>
+
+int main() {
+  sycl::queue q{};
+  auto dev = q.get_device();
+  auto ctx = q.get_context();
+  constexpr int N = 8;
+
+  if (dev.get_info<sycl::info::device::usm_shared_allocations>()) {
+    auto A = static_cast<int *>(sycl::malloc_shared(N * sizeof(int), dev, ctx));
```
Comment (bader) [inline sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp:29]:
```suggestion
  if (!dev.get_info<sycl::info::device::usm_shared_allocations>())
    return 0;
    
  auto A = static_cast<int *>(sycl::malloc_shared(N * sizeof(int), dev, ctx));
```
Please, use "early-exit" to improve readability.

File: sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// UNSUPPORTED: cuda
+// CUDA does not support unnamed lambdas.
+//
+// RUN: %clangxx -fsycl -fsyntax-only -fsycl-unnamed-lambda %s -o %t.out
```
Comment (bader) [inline sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp:4]:
```suggestion
// RUN: %clangxx -fsycl -fsyntax-only -fsycl-unnamed-lambda %s -c -o %t.out
```
I guess linking can be omitted.

File: sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
+// UNSUPPORTED: cuda
+// CUDA does not support unnamed lambdas.
```
Comment (bader) [inline sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp:2]: Can we avoid using "unnamed lambdas" to enable CUDA target?

File: sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp (hunk 1 above)
Comment (rarutyun) [reply sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp]: Done

File: sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp (hunk 2 above)
Comment (rarutyun) [reply sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp]: Done

File: sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp (hunk 3 above)
Comment (rarutyun) [reply sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp]: It tells me that -c is unknown option. Anyway, we launch this test with `syntax-only` flag that doesn't spend run-time

File: sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp (hunk 4 above)
Comment (rarutyun) [reply sycl/test/basic_tests/queue/queue_parallel_for_interface.cpp]: Yes. Done

File: sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp (hunk 1 above)
Comment (rarutyun) [reply sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp]: Actually, done from unnamed lambda perspective. But CUDA doesn't work with USM, so added XFAIL to the test

Comment (bader) [reply sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp]: I think you can leave UNSUPPORTED instead, but please, update the comment with the justification.

Comment (rarutyun) [reply sycl/test/basic_tests/queue/queue_parallel_for_generic.cpp]: Already done. Made it consistent with other bunch of tests. Justification is updated as well

**Pull request merged**